src/v/cloud_storage/topic_manifest.cc

```diff
@@ -21,12 +23,175 @@
 
 #include <boost/lexical_cast.hpp>
 #include <fmt/ostream.h>
+#include <rapidjson/encodings.h>
+#include <rapidjson/error/en.h>
 #include <rapidjson/istreamwrapper.h>
 #include <rapidjson/ostreamwrapper.h>
+#include <rapidjson/reader.h>
 
 #include <chrono>
+#include <stdexcept>
+#include <string_view>
 
 namespace cloud_storage {
+
+struct topic_manifest_handler
+  : public rapidjson::
+      BaseReaderHandler<rapidjson::UTF8<>, topic_manifest_handler> {
+    topic_manifest_handler()
+      : _state(expectManifestStart) {}
+
+    bool StartObject() {
+        vlog(cst_log.trace, "StartObject(), state_: {}", _state);
+        switch (_state) {
+        case expectManifestStart:
+            _state = expectKey;
+            return true;
+        default:
+            return false;
+        }
```

Comment:
I don't know if it's any clearer, but an option is:
```suggestion
        return std::exchange(_state, expectKey) == expectManifestStart;
```

Reply:
I'd better keep the version with `switch` for consistency